test/test_utils/GenericRequestResponseHandler.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#pragma once
+
+// #include "RSocketTests.h"
+#include "yarpl/Single.h"
+
+#include "folly/ExceptionWrapper.h"
+
+using namespace yarpl;
+using namespace yarpl::single;
+using namespace rsocket;
+using namespace rsocket::tests;
+using namespace rsocket::tests::client_server;
+
+namespace rsocket { namespace tests {
+
+using StringPair = std::pair<std::string, std::string>;
+
+inline
+std::ostream& operator<<(std::ostream& os, StringPair const& payload) {
+  return os << "('" << payload.first << "', '" << payload.second << "')";
+}
+
+struct ResponseImpl {
+  enum class Type {
+    PAYLOAD, EXCEPTION
+  };
+
+  StringPair p;
+  std::exception_ptr e;
+  Type type;
+
+  explicit ResponseImpl(StringPair const& p) :
+    p(p), type(Type::PAYLOAD) {}
+  explicit ResponseImpl(std::exception_ptr const& e) :
+    e(e), type(Type::EXCEPTION) {}
+
+  ~ResponseImpl() {}
+};
+
+using Response = std::shared_ptr<ResponseImpl>;
```

Comment:
Why does the HandlerFunc return a shared_ptr instance? 
Why don't we return a unique_ptr or the simple object (no pointer at all) and using copy elision.